calour/ms1_experiment.py

```diff
@@ -91,9 +92,102 @@ def heatmap(self, *args, **kwargs):
         return 'MS1Experiment %s with %d samples, %d features' % (
             self.description, self.data.shape[0], self.data.shape[1])
+
+    def filter_mz(self, mz, tolerance=0.001, inplace=False, negate=False):
+        '''Filter metabolites based on m/z
+
+        Parameters
+        ----------
+        mz: float or list of float
+            the M/Z to filter
+        tolerance: float, optional
+            the M/Z tolerance. filter metabolites with abs(metabolite_mz - mz) <= tolerance
+        inplace: bool, optional
+            True to replace current experiment, False to create new experiment with results
+        negate: bool, optional
+            If False, keep only metabolites matching mz
+            If True, remove metabolites matching mz
+
+        Returns
+        -------
+        calour.MS1Experiment
+            features filtered based on mz
+        '''
+        if 'MZ' not in self.feature_metadata.columns:
+            raise ValueError('The Experiment does not contain the column "MZ". cannot filter by mz')
+        mz = _to_list(mz)
+        keep = set()
+        for cmz in mz:
+            mzdiff = np.abs(self.feature_metadata['MZ'] - cmz)
+            keep = keep.union(set(np.where(mzdiff <= tolerance)[0]))
+        if negate:
+            keep = set(np.arange(len(self.feature_metadata))).difference(keep)
+        logger.info('after filtering %d features remaining' % len(keep))
+        return self.reorder(sorted(list(keep)), axis='f', inplace=inplace)
+
+    def get_bad_features(self, mz_tolerance=0.001, rt_tolerance=2, corr_thresh=0.8, inplace=False, negate=False):
+        '''Get metabolites that have similar m/z and rt, and are correlated/anti-correlated.
+        These are usually due to incorrect feature detection.
+        correlation could be due to incomplete removal of isotopes or same metabolite in multiple RTs
+        anti-correlation could be due to RT drift (splitting of one true metabolite)
+        Metabolites in the new experiment are ordered by correlation clusters
+
+        Parameters
+        ----------
+        mz_tolerance: float, optional
+            the M/Z tolerance. Metabolites are similar if abs(metabolite_mz - mz) <= mz_tolerance
+        rt_tolerance: float, optional
+            the retention time tolerance. Metabolites are similar if abs(metabolite_rt - rt) <= rt_tolerance
+        corr_threshold: float, optional
+            the minimal (abs) correlation/anti-correlation value in order to call features correlated
+        inplace: bool, optional
+            True to replace current experiment, False to create new experiment with results
+        negate: bool, optional
+            If False, keep only metabolites that show a correlation with another metabolite
+            If True, remove metabolites showing correlation
+
+        Returns
+        -------
+        calour.MS1Experiment
+            features filtered and ordered basen on m/z and rt similarity and correlation
```

Comment:
what do you do here? remove (semi-) duplicate features? or combine them?

you need a better function name!

Reply:
any ideas? this is the best name i could come up with :)

Reply:
so these are spurious duplicate features? if so, how about get_spurious_duplicates?

what's the workfflow here? what do u do after you get these features?

Reply:
feature selection in ms is complicated... and lots of parameters
the workflow is that after you set your parameters and do feature selection, you load the resulting table and look at it also using get_spurious_duplicates(). If your feature selection is too sensitive, you get lots of metabolites with similar mz/rt that show anti-correlation or correlation (depending on the exact problem).
Then, if you get too many, you go back and redo feature selection with different params...
(but you will always get some bad features - it's a sensitivity/specificity payoff... it depends on how many and how severe... and how it affects your downstream analysis)